hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CellBlocks.java

```diff
@@ -0,0 +1,427 @@
+    int index = find(k);
+    if (index >= 0) index += 1; // There's a key exactly equal.
+    else index = -(index + 1);
+    if (index < minCellIdx || index >= maxCellIdx) return null;
+    return getCellFromIndex(index);
+  }
+
+  @Override public boolean containsKey(Object o) {
+    int index = find((Cell) o);
+    return (index >= 0);
+  }
+
+  @Override public boolean containsValue(Object o) { // use containsKey(Object o) instead
+    throw new UnsupportedOperationException();
+  }
+
+  @Override public Cell get(Object o) {
+    int index = find((Cell) o);
+    if (index >= 0) {
+      return getCellFromIndex(index);
+    }
+    return null;
+  }
+
+  // -------------------------------- Entry's getters --------------------------------
+  // all interfaces returning Entries are unsupported because we are dealing only with the keys
+  @Override 
+  public Entry<Cell, Cell> lowerEntry(Cell k) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> higherEntry(Cell k) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> ceilingEntry(Cell k) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> floorEntry(Cell k) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> firstEntry() { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> lastEntry() { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> pollFirstEntry() { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public Entry<Cell, Cell> pollLastEntry() { throw new UnsupportedOperationException(); }
+
+
+  // -------------------------------- Updates --------------------------------
+  // All updating methods below are unsupported.
+  // Assuming an array of Cells will be allocated externally,
+  // fill up with Cells and provided in construction time.
+  // Later the structure is immutable.
+  @Override public Cell put(Cell k, Cell v) { throw new UnsupportedOperationException(); }
+
+  @Override public void clear() { throw new UnsupportedOperationException(); }
+
+  @Override public Cell remove(Object o) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public boolean replace(Cell k, Cell v, Cell v1) { throw new UnsupportedOperationException(); }
+
+  @Override public void putAll(Map<? extends Cell, ? extends Cell> map) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override public Cell putIfAbsent(Cell k, Cell v) { throw new UnsupportedOperationException(); }
+
+  @Override 
+  public boolean remove(Object o, Object o1) { throw new UnsupportedOperationException(); }
+
+  @Override public Cell replace(Cell k, Cell v) { throw new UnsupportedOperationException(); }
+
+
+  // -------------------------------- Sub-Sets --------------------------------
+  @Override public NavigableSet<Cell> navigableKeySet() { return new CellBlocksSet(); }
```

Comment:
this is not a sub set it should return the entire key set - but it can throw an exception - never invoked by CellSel